clip/model.py

```diff
@@ -75,50 +80,179 @@ def clip_loss(logits: mx.array) -> mx.array:
     return (caption_loss + image_loss) / 2.0
 
 
-class CLIPEncoderLayer(nn.TransformerEncoderLayer):
-    """The transformer encoder layer from CLIP."""
+class Conv2d(nn.Module):
+    def __init__(
+        self,
+        in_channels: int,
+        out_channels: int,
+        kernel_size: Union[int, tuple],
+        stride: Union[int, tuple] = 1,
+        padding: Union[int, tuple] = 0,
+        bias: bool = True,
+    ):
+        super().__init__()
 
-    def __init__(self, hidden_dim: int, intermediate_dim: int, num_heads: int):
-        super().__init__(
-            dims=hidden_dim,
-            mlp_dims=intermediate_dim,
-            num_heads=num_heads,
-            activation=quick_gelu,
-            norm_first=True,
+        kernel_size, stride, padding = map(
+            lambda x: (x, x) if isinstance(x, int) else x,
+            (kernel_size, stride, padding),
         )
+        scale = math.sqrt(1 / (in_channels * kernel_size[0] * kernel_size[1]))
+        self.weight = mx.random.uniform(
+            low=-scale,
+            high=scale,
+            shape=(out_channels, in_channels, *kernel_size),
+        )
+        if bias:
+            self.bias = mx.zeros((out_channels,))
+
+        self.padding = padding
+        self.stride = stride
+
+    def _extra_repr(self):
+        return (
+            f"{self.weight.shape[-1]}, {self.weight.shape[0]}, "
+            f"kernel_size={self.weight.shape[1:2]}, stride={self.stride}, "
+            f"padding={self.padding}, bias={'bias' in self}"
+        )
+
+    def __call__(self, x):
+        # pytorch conv2d expects the weight tensor to be of shape [out_channels, in_channels, kH, KW]
+        # mlx conv2d expects the weight tensor to be of shape [out_channels, kH, KW, in_channels]
+        y = mx.conv2d(x, self.weight.transpose(0, 2, 3, 1), self.stride, self.padding)
+        if "bias" in self:
+            y = y + self.bias
+        return y
+
+
+class CLIPAttention(nn.Module):
```

Comment:
nit, I don't think we need to prepend `CLIP` to everything in this file, maybe just the layers that are panned to export. Makes it a little verbose..